### PR TITLE
Reworking logs

### LINKED ISSUE #1732: Allow cortex logs to be forwarded to any logging service
### Description

Allow users to configure logs to be sent to any logging service (e.g. elasticsearch) instead of cloudwatch

### PR DESCRIPTION
**Pull request by vishalbollu**
closes #1732

* reduce AWS specific logging implementation in favour generic structured logging tagged with cortex specific labels
* update cortex application code to use json structured logging
* use fluent-bit to massage and stream logs to a log aggregation tool, (Cloudwatch for AWS and Stackdriver for GCP)
* update `cortex logs` CLI command to stream logs from a single running worker/replica as necessitated by the removal of AWS specific logging implementation

---

checklist:

- [ ] run `make test` and `make lint`
- [ ] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [ ] update examples
- [ ] update docs and add any new files to `summary.md` (view in [gitbook](https://cortex-labs.gitbook.io/staging/-MOmCGMADSRNQahK3Kox/) after merging)
- [ ] cherry-pick into release branches if applicable
- [ ] alert the dev team if the dev environment changed

### CODE REVIEW

File: pkg/operator/resources/validations.go [hunk 1]
```diff
@@ -141,18 +141,18 @@ func validateK8s(api *userconfig.API, virtualServices []istioclientnetworking.Vi
...
 Reserved (150 + 150) see eks.yaml for details
 */
-var _cortexCPUReserve = kresource.MustParse("710m")
+var _cortexCPUReserve = kresource.MustParse("610m")
 
 /*
 Memory Reservations:
 
-FluentD 200
+FluentBit 200
```
Comment (deliahu) [inline pkg/operator/resources/validations.go:155]: 150

File: pkg/operator/operator/workload_logging.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	_socketWriteDeadlineWait = 10 * time.Second
+	_socketCloseGracePeriod  = 10 * time.Second
+	_socketMaxMessageSize    = 8192
+	_readBufferSize          = 4096
+
+	_waitForPod = 1 * time.Second
+	_pollPeriod = 250 * time.Millisecond
+)
+
+func waitForPodPendingToComplete(podName string, podCheckCancel chan struct{}, socket *websocket.Conn) bool {
```
Comment (deliahu) [inline pkg/operator/operator/workload_logging.go:46]: `waitForPodToBeNotPending`

File: pkg/operator/operator/workload_logging.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	"github.com/gorilla/websocket"
+)
+
+const (
+	_socketWriteDeadlineWait = 10 * time.Second
+	_socketCloseGracePeriod  = 10 * time.Second
+	_socketMaxMessageSize    = 8192
+	_readBufferSize          = 4096
+
+	_waitForPod = 1 * time.Second
```
Comment (deliahu) [inline pkg/operator/operator/workload_logging.go:42]: _pendingPodCheckInterval

File: pkg/operator/operator/workload_logging.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	defer cancel()
+
+	kubectlArgs := []string{"/usr/local/bin/kubectl", "-n=" + "default", "logs", "--all-containers", podName, "--follow"}
+
+	cmd := exec.CommandContext(ctx, kubectlArgs[0], kubectlArgs[1:]...)
+
+	logStream, err := cmd.StdoutPipe()
+	if err != nil {
+		telemetry.Error(errors.ErrorUnexpected(err.Error()))
+		Logger.Error(err)
```
Comment (deliahu) [inline pkg/operator/operator/workload_logging.go:100]: Add return/cleanup

File: pkg/operator/operator/workload_logging.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	}
+	cmd.Start()
+
+	go pumpStdout(socket, logStream)
+
+waitForCancel:
+	for true {
```
Comment (deliahu) [inline pkg/operator/operator/workload_logging.go:107]: Try removing the `for true`

File: pkg/operator/operator/workload_logging.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+		select {
+		case <-podCheckCancel:
+			break waitForCancel
+		}
+	}
+
+	go func() {
```
Comment (deliahu) [inline pkg/operator/operator/workload_logging.go:114]: Add a comment explaining this part

File: pkg/operator/operator/workload_logging.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+		Logger.Error(err)
+	}
+	cmd.Start()
+
+	go pumpStdout(socket, logStream)
+
+waitForCancel:
+	for true {
+		select {
+		case <-podCheckCancel:
```
Comment (deliahu) [inline pkg/operator/operator/workload_logging.go:109]: `podCheckCancel` can be renamed

File: pkg/operator/lib/logging/logging.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	operatorZapConfig := DefaultZapConfig(operatorCortexLogLevel)
+	operatorLogger, err := operatorZapConfig.Build()
+	if err != nil {
+		panic(err)
+	}
+
+	Logger = operatorLogger.Sugar()
+}
```
Comment (miguelvr) [inline pkg/operator/lib/logging/logging.go:48]: not a huge fan of init functions... where is this called? Is there any other way we can do this?

File: pkg/operator/resources/batchapi/job.go [hunk 9]
```diff
@@ -153,38 +140,47 @@ func uploadJobSpec(jobSpec *spec.Job) error {
...
+			logger.Error("please verify that the files are not empty (the files being read can be retrieved by providing `dryRun=true` query param with your job submission")
 		}
 		errs = append(errs, setEnqueueFailedStatus(jobSpec.JobKey))
 		errs = append(errs, deleteJobRuntimeResources(jobSpec.JobKey))
 
 		err := errors.FirstError(errs...)
 		if err != nil {
 			telemetry.Error(err)
-			errors.PrintError(err)
+			logging.Logger.Error(err)
```
Comment (miguelvr) [inline pkg/operator/resources/batchapi/job.go:177]: shouldn't this be `logger.Error`? I'm a bit confused between the two different loggers purpose

File: pkg/operator/resources/batchapi/job.go [hunk 10]
```diff
@@ -207,14 +203,21 @@ func deployJob(apiSpec *spec.API, jobSpec *spec.Job, submission *schema.JobSubmi
...
+
+	logger.Error(jobErr.Error())
+	err = errors.FirstError(
 		setUnexpectedErrorStatus(jobKey),
 		deleteJobRuntimeResources(jobKey),
 	)
 	if err != nil {
 		telemetry.Error(err)
-		errors.PrintError(err)
+		logging.Logger.Error(err)
```
Comment (miguelvr) [inline pkg/operator/resources/batchapi/job.go:220]: same here: `logger` vs `logging.Logger`

**Review by miguelvr — APPROVED** [review]
Just some minor nit's and some questions regarding logging, but overall LGTM

**Pull request merged**